sched/adaptdl_sched/allocator.py

```diff
@@ -38,27 +38,100 @@ class AdaptDLAllocator(object):
     def __init__(self, expander):
         self._core_api = kubernetes.client.CoreV1Api()
         self._objs_api = kubernetes.client.CustomObjectsApi()
+        self._custom_resource = ("adaptdl.petuum.com", "v1",
+                                 "", "adaptdljobs")
         self._cluster_expander = expander
         self._policy = PolluxPolicy()
+        # lock for the two corountines in run()
+        self._lock = asyncio.Lock()
 
     async def run(self):
+        # two functionality: (1) watch for new job and start if possible.
+        # (2) periodically optimize existing jobs
+        await asyncio.gather(
+            self._allocate_one_loop(),
+            self._optimize_all_loop()
+        )
+
+    async def _allocate_one_loop(self):
+        async with kubernetes.watch.Watch() as watch:
+            while True:
+                async for event in watch.stream(
+                        self._objs_api.list_namespaced_custom_object,
+                        *self._custom_resource, timeout_seconds=60):
+                    if event["type"] == "ADDED":  # there is a n arriving job
+                        async with self._lock:
+                            await self._allocate_one(event)
+
+    async def _allocate_one(self, event):
+        # re-read the job , compared with the previous readjob
+        job = event["object"]
+        namespace = job["metadata"]["namespace"]
+        name = job["metadata"]["name"]
+
+        try:
+            job = await self._objs_api.get_namespaced_custom_object(
+                "adaptdl.petuum.com", "v1", namespace, "adaptdljobs", name
+            )
+        except kubernetes.client.rest.ApiException as exc:
+            if exc.status == 404:
+                return
+            raise  # unexpected
+
+        # some other coroutine has handled this
+        if job.get("status", {}).get("allocation") or\
+                job.get("status", {}).get("group") is not None:
```

Comment:
Hmm I feel this condition is still not very accurate, and a job could get sent through this loop multiple times (since each time the stream timeout occurs, ADDED events are generated again for every job). Maybe a better way is to let `status.allocations is None` indicate the job hasn't been seen by the allocator yet, and `status.allocations == []` indicate the job was seen by the allocator, but no nodes were allocated to it. We would only want to call `_allocate_one` for jobs with `status.allocations is None`. You'll probably also need to make sure this method will patch the job with `status.allocations == []` if it could not find an allocation for it.